### PR TITLE
Add IPsec VPN Tunnel support

### LINKED ISSUE #618: Add VPN support for NSX-T
This provider currently has only limited support for VPN (and only NSX-V).
This issue is to track work for implementing NSX-T VPN.

### LINKED ISSUE #679: vcd_nsxv_ip_set
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### Terraform Version
0.12.20
Provider 2.9.0

### Affected Resource(s)
Please list the resources as a list, for example:
vcd_nsxv_ip_set

Came across this resource whilst reading the vapp_vm docs, I used it and it worked but couldn't find it in the documentation :)

### PR DESCRIPTION
**Pull request by Didainius**
Closes #618 
Closes #679

Adds `vcd_nsxt_ipsec_vpn_tunnel` resource and data source

Additional:
* Simplifies conversion of `[]string` to `*schema.Set` by removing one hop for sets of strings
Code will become this:
```go
memberNetSet := convertStringsTotTypeSet(netIds)
```
instead of this:
```go
memberNetIds := convertToTypeSet(netIds)
memberNetSet := schema.NewSet(schema.HashSchema(&schema.Schema{Type: schema.TypeString}), memberNetIds)
```

### CODE REVIEW

File: vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+			},
+			"security_profile_customization": &schema.Schema{
+				Type:        schema.TypeList,
+				Computed:    true,
+				Description: "Security profile customization",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"ike_version": &schema.Schema{
+							Type:        schema.TypeString,
+							Required:    true,
```
Comment (dataclouder) [inline vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go:93]:
```suggestion
							Computed:    true,
```

File: vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+							Description: "Encryption algorithms to use in IPSec tunnel establishment. One of AES_128, " +
+								"AES_256, AES_GCM_128, AES_GCM_192, AES_GCM_256, NO_ENCRYPTION_AUTH_AES_GMAC_128, " +
+								"NO_ENCRYPTION_AUTH_AES_GMAC_192, NO_ENCRYPTION_AUTH_AES_GMAC_256, NO_ENCRYPTION",
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+						"tunnel_digest_algorithms": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
```
Comment (dataclouder) [inline vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go:153]:
```suggestion
							Computed: true,
```

File: vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+						"dpd_probe_internal": &schema.Schema{
+							Type:     schema.TypeInt,
+							Computed: true,
+							Description: "Value in seconds of dead probe detection interval. Minimum is 3 seconds and " +
+								"the maximum is 60 seconds",
+						},
+					},
+				},
+			},
+			// Computed attributes from here
```
Comment (dataclouder) [inline vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go:183]: Comment not needed for data source

File: vcd/resource_vcd_nsxt_app_port_profile.go [hunk 4]
```diff
@@ -375,8 +375,7 @@ func setNsxtAppPortProfileData(d *schema.ResourceData, appPortProfile *types.Nsx
 			appPortMap := make(map[string]interface{})
 			appPortMap["protocol"] = value.Protocol
 
-			destinationPortInterface := convertToTypeSet(value.DestinationPorts)
-			desinationPortSet := schema.NewSet(schema.HashSchema(&schema.Schema{Type: schema.TypeString}), destinationPortInterface)
+			desinationPortSet := convertStringsTotTypeSet(value.DestinationPorts)
```
Comment (dataclouder) [inline vcd/resource_vcd_nsxt_app_port_profile.go:378]:
```suggestion
			destinationPortSet := convertStringsTotTypeSet(value.DestinationPorts)
```

File: vcd/resource_vcd_nsxt_ipsec_vpn_tunnel.go [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+	if err != nil {
+		return diag.Errorf("error retrieving Edge Gateway: %s", err)
+	}
+
+	ipSecVpnConfig, err := nsxtEdge.GetIpSecVpnTunnelById(d.Id())
+	if err != nil {
+		if govcd.ContainsNotFound(err) {
+			d.SetId("")
+		}
+		return diag.Errorf("error retrieving NSX-T IPsec VPN Tunnel configuration for deletion: %s", err)
```
Comment (dataclouder) [inline vcd/resource_vcd_nsxt_ipsec_vpn_tunnel.go:360]:
```suggestion
		return diag.Errorf("error retrieving NSX-T IPsec VPN Tunnel configuration: %s", err)
```

**Review by dataclouder — CHANGES_REQUESTED** [review]
A few small changes required

File: vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go (hunk 1 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go]: Fixed.

File: vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go (hunk 2 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go]: Fixed.

File: vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go (hunk 3 above)
Comment (Didainius) [reply vcd/datasource_vcd_nsxt_ipsec_vpn_tunnel.go]: Fixed.

File: vcd/resource_vcd_nsxt_app_port_profile.go (hunk 4 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_app_port_profile.go]: Fixed.

File: vcd/resource_vcd_nsxt_ipsec_vpn_tunnel.go (hunk 5 above)
Comment (Didainius) [reply vcd/resource_vcd_nsxt_ipsec_vpn_tunnel.go]: Fixed.

**Review by dataclouder — APPROVED** [review]

**Pull request merged**